File: src/hosp_subset_outcome.py
import duckdb
import argparse
import logging
import pandas as pd

LOGGER = logging.getLogger(__name__)
logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - %(name)s - %(levelname)s - %(message)s'
)

def main(args): 
    """
    Example script to filter all cause hospitalization records of beneficiaries with the specified outcome. 
    """

    conn = duckdb.connect('duckpond.db')
    conn.execute(f"""
    COPY (
        SELECT inpatient.*
        FROM '{args.inpatient_prefix}_*.parquet' AS inpatient
        INNER JOIN 
            (SELECT DISTINCT bene_id
             FROM'{args.outcome_prefix}_*.parquet') AS outcome
        ON inpatient.bene_id = outcome.bene_id
    ) TO '{args.output_prefix}.parquet' 
    """)

    LOGGER.info(f"# saved filtered data at {args.output_prefix}.parquet ----")
    conn.close()


if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("--inpatient_prefix", 
                        default =  "data/input/mbsf_medpar_denom/inpatient"  
                       ) 
    parser.add_argument("--outcome_prefix", 
                        default = "data/output/medpar_outcomes/michelle_garam_00/adrd"
                       )  
    parser.add_argument("--output_prefix", 
                    default = "data/output/medpar_outcomes/michelle_garam_00/hosp_subset_adrd"
                   )
    args = parser.parse_args()
    
    main(args)

File: hosp_subset_outcome.R
# Install required packages 
#install.packages("duckdb")
#install.packages("glue")

# Load required libraries
library(duckdb)
library(glue)

args <- commandArgs(trailingOnly = TRUE)

# prefixes 
inpatient_prefix <- "data/input/mbsf_medpar_denom/inpatient"
outcome_prefix <- "data/output/medpar_outcomes/michelle_garam_00/adrd"
output_prefix <- "data/output/medpar_outcomes/michelle_garam_00/hosp_subset_adrd"

if (length(args) >= 1) inpatient_prefix <- args[1]
if (length(args) >= 2) outcome_prefix <- args[2]
if (length(args) >= 3) output_prefix <- args[3]

# Main function
main <- function(inpatient_prefix, outcome_prefix, output_prefix) {
  # Connect to DuckDB 
  con <- dbConnect(duckdb::duckdb(), dbdir = "duckpond.db", read_only = FALSE)
  query <- glue("
    COPY (
      SELECT inpatient.*
      FROM '{inpatient_prefix}_*.parquet' AS inpatient
      INNER JOIN (
        SELECT DISTINCT bene_id
        FROM '{outcome_prefix}_*.parquet'
      ) AS outcome
      ON inpatient.bene_id = outcome.bene_id
    ) TO '{output_prefix}.parquet'
  ")
  
  # Execute the query and save the result
  tryCatch({
    dbExecute(con, query)
    message(sprintf("Filtered data saved at: %s.parquet", output_prefix))
  }, error = function(e) {
    stop("Error during query execution: ", e$message)
  }, finally = {
    dbDisconnect(con, shutdown = TRUE)
  })
}

# Execute the main function
main(inpatient_prefix, outcome_prefix, output_prefix)